crates/fluvio/src/consumer/retry.rs

```diff
@@ -307,52 +292,113 @@ impl ConsumerRetryStream {
             backoff_and_wait(&mut backoff).await;
             attempts += 1;
 
-            // Determine the offset for reconnection.
-            let offset = if let Some(next) = inner.next_offset_to_read {
-                match Offset::absolute(next) {
-                    Ok(off) => off,
-                    Err(e) => {
-                        warn!(target: SPAN_RETRY, "Error creating offset: {}", e);
-                        return (stream.clone(), Some(Err(ErrorCode::OffsetOutOfRange)));
-                    }
-                }
-            } else {
-                inner.consumer_config.offset_start.clone()
-            };
-
             // Update the consumer configuration with the new offset.
             let mut new_config = inner.consumer_config.clone();
-            new_config.offset_start = offset;
+            new_config.offset_start = inner.next_offset();
 
-            // Reconnect loop: keep trying until a new stream is created.
-            loop {
-                // match Self::reconnect_stream(&inner, new_config.clone(), backoff.clone()).await {
-                match strategy
-                    .reconnect(&inner, new_config.clone(), backoff.clone())
-                    .await
-                {
-                    Ok(new_stream) => {
-                        info!(target: SPAN_RETRY, "Created new consumer stream with offset: {:?}", new_config.offset_start);
-                        stream = new_stream;
-                        break;
+            // Reconnect loop
+            match Self::handle_reconnection_loop(&mut inner, backoff.clone(), attempts, new_config)
+                .await
+            {
+                Ok(new_stream) => {
+                    drop(stream); // Drop the old stream
+                    stream = new_stream;
+                }
+                Err(err) => {
+                    return (stream.clone(), Some(Err(err)));
+                }
+            }
+        }
+    }
+
+    /// Handles the consumer records stream.
+    async fn handle_consumer_records(
+        inner: &mut ConsumerRetryInner,
+        stream: ShararedConsumerStream,
+    ) -> Option<ConsumerFutureOutput> {
+        let mut stream_lock = stream.lock().await;
```

Comment:
locking is expensive.  what's performance drop here?

Reply:
Before we were using lock already, I changed to Optional with `take` to borrow, but this is not too safe, and also it makes really hard to test.